app/src/main/java/com/github/gotify/api/CertUtils.java

```diff
@@ -60,21 +70,62 @@ public static void applySslSettings(OkHttpClient.Builder builder, SSLSettings se
             }
 
             if (settings.cert != null) {
+                KeyManager[] keyManagers = new KeyManager[] {};
                 TrustManager[] trustManagers = certToTrustManager(settings.cert);
 
                 if (trustManagers != null && trustManagers.length > 0) {
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                        if (settings.clientCert != null) {
+                            keyManagers = getClientCerts(settings.clientCert, settings.clientCertPassword.toCharArray());
+                        }
+                    }
+
                     SSLContext context = SSLContext.getInstance("TLS");
-                    context.init(new KeyManager[] {}, trustManagers, new SecureRandom());
+                    context.init(keyManagers, trustManagers, new SecureRandom());
                     builder.sslSocketFactory(
                             context.getSocketFactory(), (X509TrustManager) trustManagers[0]);
                 }
+            } else {
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                    if (settings.clientCert != null) {
+                        KeyManager[] keyManagers = getClientCerts(settings.clientCert, settings.clientCertPassword.toCharArray());
+
+                        TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(
+                                TrustManagerFactory.getDefaultAlgorithm());
+                        trustManagerFactory.init((KeyStore) null);
+                        TrustManager[] trustManagers = trustManagerFactory.getTrustManagers();
+                        if (trustManagers.length != 1 || !(trustManagers[0] instanceof X509TrustManager)) {
+                            throw new IllegalStateException("Unexpected default trust managers:");
+                        }
+                        X509TrustManager trustManager = (X509TrustManager) trustManagers[0];
```

Comment:
Why is this check needed, couldn't we pass all the trust managers into `context.init(keyManagers, trustManagers, null);`?

Reply:
I believe that was in some of the code I used as reference. I will check those calls and if there's always a default trust manager, we could remove it.